ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBlobCryptoAgentImpl.java

```diff
@@ -0,0 +1,163 @@
+   * |(2 bytes)|  Key Size    |  Data Size      | Key       | Data     |(8 bytes) |
+   * |         |              |                 |           | (m bytes)|          |
+   * |         |  (4 bytes)   | (8 bytes)       | (n bytes) |          |          |
+   * |         |              |                 |           |          |          |
+   *  - - - - - - - - - - - - - - - - - - -- - -- - - - - - -- - - - - -- - - - - -
+   *
+   *  version         - The version of the message header
+   *
+   *  encrypted key   - The size of the encrypted key (should be >= 0).
+   *  size
+   *
+   *  encrypted data  - The size of the encrypted data (should be >= 0).
+   *  size              8 bytes for future proofing, but current implementation only
+   *                    supports int sized sizes
+   *
+   *  encrypted key   - The encrypted key
+   *
+   *  encrypted data  - The encrypted data
+   *
+   *  crc             - The crc of the message
+   *
+   */
+  static class EncryptedKeyAndEncryptedData {
+
+    private static short CURRENT_VERSION = 1;
+    private static int VERSION_FIELD_SIZE = 2;
+    private static int ENCRYPTED_KEY_SIZE_FIELD_SIZE = 4;
+    private static int ENCRYPTED_DATA_SIZE_FIELD_SIZE = 8;
+    static int INITIAL_MESSAGE_LENGTH =
+        VERSION_FIELD_SIZE + ENCRYPTED_KEY_SIZE_FIELD_SIZE + ENCRYPTED_DATA_SIZE_FIELD_SIZE;
+    private static int CRC_FIELD_LENGTH = 8;
+    static int OVERHEAD_LENGTH = INITIAL_MESSAGE_LENGTH + CRC_FIELD_LENGTH;
+
+    private ByteBuffer encryptedKey;
+    private ByteBuffer encryptedData;
+
+    EncryptedKeyAndEncryptedData(ByteBuffer encryptedKey, ByteBuffer encryptedData) {
+      this.encryptedKey = encryptedKey;
+      this.encryptedData = encryptedData;
+    }
+
+    static void serialize(ByteBuffer outputBuffer, EncryptedKeyAndEncryptedData encryptedKeyAndEncryptedData) {
+      int startOffset = outputBuffer.position();
+      outputBuffer.putShort(CURRENT_VERSION);
+      int encryptedKeySize = encryptedKeyAndEncryptedData.encryptedKey.array().length;
+      int encryptedDataSize = encryptedKeyAndEncryptedData.encryptedData.array().length;
+      outputBuffer.putInt(encryptedKeySize);
+      outputBuffer.putLong(encryptedDataSize);
+      outputBuffer.put(encryptedKeyAndEncryptedData.encryptedKey);
+      outputBuffer.put(encryptedKeyAndEncryptedData.encryptedData);
+      Crc32 crc = new Crc32();
+      crc.update(outputBuffer.array(), startOffset, INITIAL_MESSAGE_LENGTH + encryptedKeySize + encryptedDataSize);
+      outputBuffer.putLong(crc.getValue());
+    }
+
+    static EncryptedKeyAndEncryptedData deserialize(ByteBuffer buffer) throws GeneralSecurityException {
+      int startOffset = buffer.position();
+      Crc32 crc = new Crc32();
+      short version = buffer.getShort();
+      if (version != 1) {
+        throw new GeneralSecurityException("Unrecognized version");
+      }
+      int encryptedKeySize = buffer.getInt();
```

Comment:
Should we add specified (de)serialization based on version?

if version == 1{

} else {
....
}

as a placeholder.

Reply:
Original intent was to add this when we actually had new versions